demos/conditionvariables/README.md

```diff
@@ -0,0 +1,50 @@
+# Condition Variables
+
+In this demo we explain the mechanism in condition variables.
+
+Condition variable is a synchronization tool which can block threads until a certain condition happens. 
+
+To show the concept behind it, we need to introduce the following:
+
+* `init_condition_variable` (in `src/user/condition_variable.c`)
+* `wait` (in `src/user/condition_variable.c`)
+* `signal` (in `src/user/condition_variable.c`)
+* `broadcast` (in `src/user/condition_variable.c`)
+
+## `init_condition_variable`
+
+This function initializes the condition variable pointed by `cv`.
+
+Before using functions like `wait()`, `signal()` and `broadcast()`, we must set up the variables with this function.
+
+## `wait`
+
+When we use this function, the thread will be blocked by the condition variable pointed by `cv`.
+
+After calling the function, the thread's `ThreadState` will be changed to `wait`. Since the scheduler only chooses the thread whose state is either `init`, `running` or `suspended`, the blocking thread will keep waiting until other threads signal through the same condition variable.
+
+## `signal`
+
+This function unblocks a thread that is blocked on the specified condition variable `cv`.
+
+`signal()` unlocks the thread by setting the state to `suspended`. So the thread can be chosen by the scheduler.
+
+If there exists a thread that can be waked up, the function returns `true`. When there has no thread can be signalled, it returns `false` as a consequence.
```

Comment:
The meaning here is fine but some phrasing changes would help.

"If there was a waiting thread that was signalled, the function returns `true`. When there are no threads that can be signalled, it returns `false`."

Side note: it would be "woken up" but I admit it also took me a while to work that one out :)